lib/lints/avoid_unnecessary_type_assertions/avoid_unnecessary_type_assertions_rule.dart

```diff
@@ -0,0 +1,183 @@
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:analyzer/error/error.dart';
+import 'package:analyzer/error/listener.dart';
+import 'package:analyzer/source/source_range.dart';
+import 'package:collection/collection.dart';
+import 'package:custom_lint_builder/custom_lint_builder.dart';
+import 'package:solid_lints/models/rule_config.dart';
+import 'package:solid_lints/models/solid_lint_rule.dart';
+import 'package:solid_lints/utils/types_utils.dart';
+
+part 'avoid_unnecessary_type_assertions_fix.dart';
+
+/// A `avoid-unnecessary-type-assertions` rule which
+/// warns about unnecessary usage of `is` and `whereType` operators
+
+class AvoidUnnecessaryTypeAssertions extends SolidLintRule {
+  /// The [LintCode] of this lint rule that represents
+  /// the error whether we use bad formatted double literals.
+  static const lintName = 'avoid-unnecessary-type-assertions';
+
+  static const _unnecessaryIsCode = LintCode(
+    name: lintName,
+    problemMessage: "Unnecessary usage of the 'is' operator.",
+  );
+
+  static const _unnecessaryWhereTypeCode = LintCode(
+    name: lintName,
+    problemMessage: "Unnecessary usage of the 'whereType'.",
+  );
+
+  AvoidUnnecessaryTypeAssertions._(super.config);
+
+  /// Creates a new instance of [AvoidUnnecessaryTypeAssertions]
+  /// based on the lint configuration.
+  factory AvoidUnnecessaryTypeAssertions.createRule(CustomLintConfigs configs) {
+    final rule = RuleConfig(
+      configs: configs,
+      name: lintName,
+      problemMessage: (_) =>
+          "Unnecessary usage of 'is' or 'whereType' operators.",
```

Comment:
Generalize. Especially given we're going to expand this rule with `case` statements support later on.
```suggestion
          "Unnecessary usage of typecast operators.",
```